src/main/telemetry/msp_shared.c

```diff
@@ -29,225 +29,294 @@
+                // packet loss detected!
+                mspStarted = 0;
+                return false;
+            }
+        }
+        sbufInit(&sbufInput, crsfPayload + 1, crsfPayload + crsfPayloadLength);
     }
 
-    const uint8_t bufferBytesRemaining = sbufBytesRemaining(rxBuf);
-    const uint8_t frameBytesRemaining = sbufBytesRemaining(frameBuf);
-    uint8_t payload[frameBytesRemaining];
+    lastSeq = seqNumber;
 
-    if (bufferBytesRemaining >= frameBytesRemaining) {
-        sbufReadData(frameBuf, payload, frameBytesRemaining);
-        sbufAdvance(frameBuf, frameBytesRemaining);
-        sbufWriteData(rxBuf, payload, frameBytesRemaining);
-        lastSeq = seqNumber;
+    const int payloadExpecting = sbufBytesRemaining(&requestPacket.buf);
+    const int payloadIncoming = sbufBytesRemaining(&sbufInput);
 
+    if (payloadExpecting > payloadIncoming) {
+        sbufWriteData(&requestPacket.buf, sbufInput.ptr, payloadIncoming);
+        sbufAdvance(&sbufInput, payloadIncoming);
         return false;
-    } else {
-        sbufReadData(frameBuf, payload, bufferBytesRemaining);
-        sbufAdvance(frameBuf, bufferBytesRemaining);
-        sbufWriteData(rxBuf, payload, bufferBytesRemaining);
-        sbufSwitchToReader(rxBuf, mspPackage.requestBuffer);
-        while (sbufBytesRemaining(rxBuf)) {
-            checksum ^= sbufReadU8(rxBuf);
-        }
-
-        if (checksum != *frameBuf->ptr) {
-            mspStarted = 0;
-            sendMspErrorResponse(TELEMETRY_MSP_CRC_ERROR, packet->cmd);
-            return true;
+    } else { // this is the last/only chunk 
+        if (payloadExpecting) {
+            sbufWriteData(&requestPacket.buf, sbufInput.ptr, payloadExpecting);
+            sbufAdvance(&sbufInput, payloadExpecting);
         }
     }
 
     // Skip a few telemetry requests if command is MSP_EEPROM_WRITE
-    if (packet->cmd == MSP_EEPROM_WRITE && skipsBeforeResponse) {
+    if (requestPacket.cmd == MSP_EEPROM_WRITE && skipsBeforeResponse) {
         *skipsBeforeResponse = TELEMETRY_REQUEST_SKIPS_AFTER_EEPROMWRITE;
     }
 
     mspStarted = 0;
-    sbufSwitchToReader(rxBuf, mspPackage.requestBuffer);
+    sbufSwitchToReader(&requestPacket.buf, requestBuffer);
     processMspPacket();
     return true;
 }
 
-bool sendMspReply(uint8_t payloadSize, mspResponseFnPtr responseFn)
+bool sendMspReply(const uint8_t payloadSizeMax, mspResponseFnPtr responseFn)
 {
-    static uint8_t checksum = 0;
     static uint8_t seq = 0;
 
-    uint8_t payloadOut[payloadSize];
+    uint8_t crsfPayloadBuf[payloadSizeMax];
     sbuf_t payload;
-    sbuf_t *payloadBuf = sbufInit(&payload, payloadOut, payloadOut + payloadSize);
-    sbuf_t *txBuf = &mspPackage.responsePacket->buf;
+    sbuf_t *payloadBuf = sbufInit(&payload, crsfPayloadBuf, crsfPayloadBuf + payloadSizeMax);
+    const uint8_t mspVersion = responsePacket.cmd < 0xff ? 1 : 2; // command ID higher that 0xff may be sent only as MSPv2
```

Comment:
I still think we should respond with the same version as the request.

Reply:
I started implementing it and faced... well, not a problem but an eager to change more and more.
The best place to save version of request is the mspPacket_s structure. But then the structure is used heavily in msp_serial module and it'll be unfinished if not using the field in the module too. It'll look messy if there is a field for version but e.g., mspSerialEncode() is using an mspVersion parameter, ignoring the field in the structure.
And it only seems reasonable to make the logic of MSP communication identical on serial and telemetry. 
So, I don't mind implementing this feature but maybe it's not worth it putting so many changes in one PR? I think it's better to implement it in another PR.
Of course it's possible to make a temporary solution and save the request version in msp_shared module. But it seems messy and I don't think it worth it.
I found some other issues in this mechanism that I'd like to improve but I think it's better to keep amount of changes in a PR at minimum.

Reply:
I agree that the mspPacket structure would be the best place for it. 
I also agree that we shouldn't change to many other things not related to the functionality in this PR.

> Of course it's possible to make a temporary solution and save the request version in msp_shared module. But it seems messy and I don't think it worth it.

I don't think it would be more messy than having one behaviour in 4.3 and a slightly different behaviour in 4.4. IMO storing the msp version locally in msp_shared is acceptable and will give us the desired behaviour.
What's important here is that we can agree on how we want this to work in this PR and not have to make incremental changes afterwards that will have to be accounted for by the client. Changing HOW it's implemented afterwards is of course perfectly fine.

Reply:
Ok. It's done locally in msp_shared module.